// @colyseus/tools/build/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/tools/src/index.ts
var index_exports = {};
__export(index_exports, {
  default: () => index_default,
  getTransport: () => getTransport,
  listen: () => listen
});
module.exports = __toCommonJS(index_exports);
var import_loadenv = require("./loadenv.cjs");
var import_os = __toESM(require("os"), 1);
var import_fs = __toESM(require("fs"), 1);
var import_net = __toESM(require("net"), 1);
var import_http = __toESM(require("http"), 1);
var import_core = require("@colyseus/core");
var import_ws_transport = require("@colyseus/ws-transport");
var BunWebSockets = (0, import_core.dynamicImport)("@colyseus/bun-websockets");
var RedisDriver = (0, import_core.dynamicImport)("@colyseus/redis-driver");
var RedisPresence = (0, import_core.dynamicImport)("@colyseus/redis-presence");
var ALLOWED_KEYS = {
  "displayLogs": "boolean",
  "options": "object",
  "rooms": "object",
  "routes": "object",
  "initializeTransport": "function",
  "initializeExpress": "function",
  "initializeGameServer": "function",
  "beforeListen": "function",
  // deprecated options (will be removed in the next major version)
  "getId": "function"
};
function index_default(options) {
  for (const option in options) {
    if (!ALLOWED_KEYS[option]) {
      throw new Error(`\u274C Invalid option '${option}'. Allowed options are: ${Object.keys(ALLOWED_KEYS).join(", ")}`);
    }
    if (options[option] !== void 0 && typeof options[option] !== ALLOWED_KEYS[option]) {
      throw new Error(`\u274C Invalid type for ${option}: please provide a ${ALLOWED_KEYS[option]} value.`);
    }
  }
  return options;
}
async function listen(options, port = Number(process.env.PORT || 2567)) {
  if (process.env.COLYSEUS_CLOUD !== void 0) {
    port = 2567;
  }
  const processNumber = Number(process.env.NODE_APP_INSTANCE || "0");
  port += processNumber;
  let server;
  let displayLogs = true;
  if (options instanceof import_core.Server) {
    server = options;
  } else {
    server = await buildServerFromOptions(options, port);
    displayLogs = options.displayLogs;
    await options.initializeGameServer?.(server);
    await import_core.matchMaker.onReady;
    await options.beforeListen?.();
  }
  if (process.env.COLYSEUS_CLOUD !== void 0) {
    const socketPath = `/run/colyseus/${port}.sock`;
    await checkInactiveSocketFile(socketPath);
    await server.listen(
      socketPath,
      0
      // workaround to allow using @colyseus/core's .listen() directly on Colyseus Cloud
    );
  } else {
    await server.listen(port);
  }
  if (typeof process.send === "function") {
    process.send("ready");
  }
  if (displayLogs) {
    import_core.logger.info(`\u2694\uFE0F  Listening on http://localhost:${port}`);
  }
  return server;
}
async function buildServerFromOptions(options, port) {
  const serverOptions = options.options || {};
  options.displayLogs = options.displayLogs ?? true;
  if (process.env.COLYSEUS_CLOUD !== void 0) {
    const cloudConfig = await getColyseusCloudConfig(port, serverOptions.driver, serverOptions.presence);
    if (cloudConfig) {
      serverOptions.driver = cloudConfig.driver;
      serverOptions.presence = cloudConfig.presence;
      serverOptions.publicAddress = cloudConfig.publicAddress;
    }
  }
  return (0, import_core.defineServer)({
    rooms: options.rooms || {},
    routes: options.routes,
    ...serverOptions,
    express: options.initializeExpress,
    transport: await getTransport(options)
  });
}
async function getTransport(options) {
  let transport;
  if (!options.initializeTransport) {
    if (typeof Bun !== "undefined") {
      BunWebSockets.catch(() => {
        import_core.logger.warn("");
        import_core.logger.warn("\u274C could not initialize BunWebSockets.");
        import_core.logger.warn("\u{1F449} npm install --save @colyseus/bun-websockets");
        import_core.logger.warn("");
      });
      const module2 = await BunWebSockets;
      options.initializeTransport = (options2) => new module2.BunWebSockets(options2);
    } else {
      options.initializeTransport = (options2) => new import_ws_transport.WebSocketTransport(options2);
    }
  }
  const server = import_http.default.createServer();
  transport = await options.initializeTransport({ server });
  return transport;
}
async function getColyseusCloudConfig(port, currentDriver, currentPresence) {
  const useRedisConfig = import_os.default.cpus().length > 1 || process.env.REDIS_URI !== void 0;
  if (!useRedisConfig) {
    return null;
  }
  let driver = currentDriver;
  let presence = currentPresence;
  const publicAddress = process.env.SUBDOMAIN + "." + process.env.SERVER_NAME + "/" + port;
  if (!driver) {
    try {
      const module2 = await RedisDriver;
      driver = new module2.RedisDriver(process.env.REDIS_URI);
    } catch (e) {
      console.error(e);
      import_core.logger.warn("");
      import_core.logger.warn("\u274C could not initialize RedisDriver.");
      import_core.logger.warn("\u{1F449} npm install --save @colyseus/redis-driver");
      import_core.logger.warn("");
    }
  }
  if (!presence) {
    try {
      const module2 = await RedisPresence;
      presence = new module2.RedisPresence(process.env.REDIS_URI);
    } catch (e) {
      console.error(e);
      import_core.logger.warn("");
      import_core.logger.warn("\u274C could not initialize RedisPresence.");
      import_core.logger.warn("\u{1F449} npm install --save @colyseus/redis-presence");
      import_core.logger.warn("");
    }
  }
  return { driver, presence, publicAddress };
}
function checkInactiveSocketFile(sockFilePath) {
  return new Promise((resolve, reject) => {
    const client = import_net.default.createConnection({ path: sockFilePath }).on("connect", () => {
      client.end();
      throw new Error(`EADDRINUSE: Already listening on '${sockFilePath}'`);
    }).on("error", () => {
      import_fs.default.unlink(sockFilePath, () => resolve(true));
    });
  });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getTransport,
  listen
});
