// @colyseus/tools/build/index.mjs
// packages/tools/src/index.ts
import "./loadenv.mjs";
import os from "os";
import fs from "fs";
import net from "net";
import http from "http";
import {
  logger,
  Server,
  matchMaker,
  defineServer,
  dynamicImport
} from "@colyseus/core";
import { WebSocketTransport } from "@colyseus/ws-transport";
var BunWebSockets = dynamicImport("@colyseus/bun-websockets");
var RedisDriver = dynamicImport("@colyseus/redis-driver");
var RedisPresence = dynamicImport("@colyseus/redis-presence");
var ALLOWED_KEYS = {
  "displayLogs": "boolean",
  "options": "object",
  "rooms": "object",
  "routes": "object",
  "initializeTransport": "function",
  "initializeExpress": "function",
  "initializeGameServer": "function",
  "beforeListen": "function",
  // deprecated options (will be removed in the next major version)
  "getId": "function"
};
function index_default(options) {
  for (const option in options) {
    if (!ALLOWED_KEYS[option]) {
      throw new Error(`\u274C Invalid option '${option}'. Allowed options are: ${Object.keys(ALLOWED_KEYS).join(", ")}`);
    }
    if (options[option] !== void 0 && typeof options[option] !== ALLOWED_KEYS[option]) {
      throw new Error(`\u274C Invalid type for ${option}: please provide a ${ALLOWED_KEYS[option]} value.`);
    }
  }
  return options;
}
async function listen(options, port = Number(process.env.PORT || 2567)) {
  if (process.env.COLYSEUS_CLOUD !== void 0) {
    port = 2567;
  }
  const processNumber = Number(process.env.NODE_APP_INSTANCE || "0");
  port += processNumber;
  let server;
  let displayLogs = true;
  if (options instanceof Server) {
    server = options;
  } else {
    server = await buildServerFromOptions(options, port);
    displayLogs = options.displayLogs;
    await options.initializeGameServer?.(server);
    await matchMaker.onReady;
    await options.beforeListen?.();
  }
  if (process.env.COLYSEUS_CLOUD !== void 0) {
    const socketPath = `/run/colyseus/${port}.sock`;
    await checkInactiveSocketFile(socketPath);
    await server.listen(
      socketPath,
      0
      // workaround to allow using @colyseus/core's .listen() directly on Colyseus Cloud
    );
  } else {
    await server.listen(port);
  }
  if (typeof process.send === "function") {
    process.send("ready");
  }
  if (displayLogs) {
    logger.info(`\u2694\uFE0F  Listening on http://localhost:${port}`);
  }
  return server;
}
async function buildServerFromOptions(options, port) {
  const serverOptions = options.options || {};
  options.displayLogs = options.displayLogs ?? true;
  if (process.env.COLYSEUS_CLOUD !== void 0) {
    const cloudConfig = await getColyseusCloudConfig(port, serverOptions.driver, serverOptions.presence);
    if (cloudConfig) {
      serverOptions.driver = cloudConfig.driver;
      serverOptions.presence = cloudConfig.presence;
      serverOptions.publicAddress = cloudConfig.publicAddress;
    }
  }
  return defineServer({
    rooms: options.rooms || {},
    routes: options.routes,
    ...serverOptions,
    express: options.initializeExpress,
    transport: await getTransport(options)
  });
}
async function getTransport(options) {
  let transport;
  if (!options.initializeTransport) {
    if (typeof Bun !== "undefined") {
      BunWebSockets.catch(() => {
        logger.warn("");
        logger.warn("\u274C could not initialize BunWebSockets.");
        logger.warn("\u{1F449} npm install --save @colyseus/bun-websockets");
        logger.warn("");
      });
      const module = await BunWebSockets;
      options.initializeTransport = (options2) => new module.BunWebSockets(options2);
    } else {
      options.initializeTransport = (options2) => new WebSocketTransport(options2);
    }
  }
  const server = http.createServer();
  transport = await options.initializeTransport({ server });
  return transport;
}
async function getColyseusCloudConfig(port, currentDriver, currentPresence) {
  const useRedisConfig = os.cpus().length > 1 || process.env.REDIS_URI !== void 0;
  if (!useRedisConfig) {
    return null;
  }
  let driver = currentDriver;
  let presence = currentPresence;
  const publicAddress = process.env.SUBDOMAIN + "." + process.env.SERVER_NAME + "/" + port;
  if (!driver) {
    try {
      const module = await RedisDriver;
      driver = new module.RedisDriver(process.env.REDIS_URI);
    } catch (e) {
      console.error(e);
      logger.warn("");
      logger.warn("\u274C could not initialize RedisDriver.");
      logger.warn("\u{1F449} npm install --save @colyseus/redis-driver");
      logger.warn("");
    }
  }
  if (!presence) {
    try {
      const module = await RedisPresence;
      presence = new module.RedisPresence(process.env.REDIS_URI);
    } catch (e) {
      console.error(e);
      logger.warn("");
      logger.warn("\u274C could not initialize RedisPresence.");
      logger.warn("\u{1F449} npm install --save @colyseus/redis-presence");
      logger.warn("");
    }
  }
  return { driver, presence, publicAddress };
}
function checkInactiveSocketFile(sockFilePath) {
  return new Promise((resolve, reject) => {
    const client = net.createConnection({ path: sockFilePath }).on("connect", () => {
      client.end();
      throw new Error(`EADDRINUSE: Already listening on '${sockFilePath}'`);
    }).on("error", () => {
      fs.unlink(sockFilePath, () => resolve(true));
    });
  });
}
export {
  index_default as default,
  getTransport,
  listen
};
