// @colyseus/core/build/Debug.mjs
// packages/core/src/Debug.ts
import debug from "debug";
import { logger } from "./Logger.mjs";
import { ServerError } from "./errors/ServerError.mjs";
var debugConnection = debug("colyseus:connection");
debugConnection.log = console.debug.bind(console);
var debugDriver = debug("colyseus:driver");
debugDriver.log = console.debug.bind(console);
var debugMatchMaking = debug("colyseus:matchmaking");
debugMatchMaking.log = console.debug.bind(console);
var debugMessage = debug("colyseus:message");
debugMessage.log = console.debug.bind(console);
var debugPatch = debug("colyseus:patch");
debugPatch.log = console.debug.bind(console);
var debugPresence = debug("colyseus:presence");
debugPresence.log = console.debug.bind(console);
var debugError = debug("colyseus:errors");
debugError.log = console.error.bind(console);
var debugDevMode = debug("colyseus:devmode");
debugDevMode.log = console.debug.bind(console);
var debugAndPrintError = (e) => {
  const message = e instanceof Error ? e.stack : e;
  if (!(e instanceof ServerError)) {
    logger.error(message);
  }
  debugError.call(debugError, message);
};
export {
  debugAndPrintError,
  debugConnection,
  debugDevMode,
  debugDriver,
  debugError,
  debugMatchMaking,
  debugMessage,
  debugPatch,
  debugPresence
};
