// @colyseus/core/build/Debug.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/core/src/Debug.ts
var Debug_exports = {};
__export(Debug_exports, {
  debugAndPrintError: () => debugAndPrintError,
  debugConnection: () => debugConnection,
  debugDevMode: () => debugDevMode,
  debugDriver: () => debugDriver,
  debugError: () => debugError,
  debugMatchMaking: () => debugMatchMaking,
  debugMessage: () => debugMessage,
  debugPatch: () => debugPatch,
  debugPresence: () => debugPresence
});
module.exports = __toCommonJS(Debug_exports);
var import_debug = __toESM(require("debug"), 1);
var import_Logger = require("./Logger.cjs");
var import_ServerError = require("./errors/ServerError.cjs");
var debugConnection = (0, import_debug.default)("colyseus:connection");
debugConnection.log = console.debug.bind(console);
var debugDriver = (0, import_debug.default)("colyseus:driver");
debugDriver.log = console.debug.bind(console);
var debugMatchMaking = (0, import_debug.default)("colyseus:matchmaking");
debugMatchMaking.log = console.debug.bind(console);
var debugMessage = (0, import_debug.default)("colyseus:message");
debugMessage.log = console.debug.bind(console);
var debugPatch = (0, import_debug.default)("colyseus:patch");
debugPatch.log = console.debug.bind(console);
var debugPresence = (0, import_debug.default)("colyseus:presence");
debugPresence.log = console.debug.bind(console);
var debugError = (0, import_debug.default)("colyseus:errors");
debugError.log = console.error.bind(console);
var debugDevMode = (0, import_debug.default)("colyseus:devmode");
debugDevMode.log = console.debug.bind(console);
var debugAndPrintError = (e) => {
  const message = e instanceof Error ? e.stack : e;
  if (!(e instanceof import_ServerError.ServerError)) {
    import_Logger.logger.error(message);
  }
  debugError.call(debugError, message);
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  debugAndPrintError,
  debugConnection,
  debugDevMode,
  debugDriver,
  debugError,
  debugMatchMaking,
  debugMessage,
  debugPatch,
  debugPresence
});
