// @colyseus/core/build/index.mjs
// packages/core/src/index.ts
import { ClockTimer as Clock, Delayed } from "@colyseus/timer";
import {
  Protocol,
  ErrorCode,
  CloseCode
} from "@colyseus/shared-types";
import { Server, defineRoom, defineServer, registerRoomDefinitions, unregisterRoomDefinitions } from "./Server.mjs";
import { Room, room, RoomInternalState, validate } from "./Room.mjs";
import { getMessageBytes } from "./Protocol.mjs";
import { RegisteredHandler } from "./matchmaker/RegisteredHandler.mjs";
import { ServerError } from "./errors/ServerError.mjs";
import {
  OnCreateException,
  OnAuthException,
  OnJoinException,
  OnLeaveException,
  OnDisposeException,
  OnMessageException,
  SimulationIntervalException,
  TimedEventException
} from "./errors/RoomExceptions.mjs";
import * as matchMaker from "./MatchMaker.mjs";
import { updateLobby, subscribeLobby } from "./matchmaker/Lobby.mjs";
import { createNodeMatchmakingMiddleware } from "./router/node.mjs";
export * from "./matchmaker/LocalDriver/LocalDriver.mjs";
import { initializeRoomCache } from "./matchmaker/driver.mjs";
import { ClientState, ClientArray, Transport, connectClientToRoom } from "./Transport.mjs";
import {} from "./presence/Presence.mjs";
import { LocalPresence } from "./presence/LocalPresence.mjs";
import {} from "./serializer/Serializer.mjs";
import { SchemaSerializer } from "./serializer/SchemaSerializer.mjs";
import { generateId, Deferred, spliceOne, getBearerToken, dynamicImport } from "./utils/Utils.mjs";
import { isDevMode, setDevMode } from "./utils/DevMode.mjs";
import { subscribeIPC, requestFromIPC } from "./IPC.mjs";
import {
  debugMatchMaking,
  debugMessage,
  debugPatch,
  debugError,
  debugConnection,
  debugDriver,
  debugPresence,
  debugAndPrintError
} from "./Debug.mjs";
import { LobbyRoom } from "./rooms/LobbyRoom.mjs";
import { RelayRoom } from "./rooms/RelayRoom.mjs";
import { QueueRoom } from "./rooms/QueueRoom.mjs";
import {
  createEndpoint,
  createInternalContext,
  createMiddleware,
  createRouter,
  toNodeHandler,
  __globalEndpoints
} from "./router/index.mjs";
import { logger } from "./Logger.mjs";
export {
  ClientArray,
  ClientState,
  Clock,
  CloseCode,
  Deferred,
  Delayed,
  ErrorCode,
  LobbyRoom,
  LocalPresence,
  OnAuthException,
  OnCreateException,
  OnDisposeException,
  OnJoinException,
  OnLeaveException,
  OnMessageException,
  Protocol,
  QueueRoom,
  RegisteredHandler,
  RelayRoom,
  Room,
  RoomInternalState,
  SchemaSerializer,
  Server,
  ServerError,
  SimulationIntervalException,
  TimedEventException,
  Transport,
  __globalEndpoints,
  connectClientToRoom,
  createEndpoint,
  createInternalContext,
  createMiddleware,
  createNodeMatchmakingMiddleware,
  createRouter,
  debugAndPrintError,
  debugConnection,
  debugDriver,
  debugError,
  debugMatchMaking,
  debugMessage,
  debugPatch,
  debugPresence,
  defineRoom,
  defineServer,
  dynamicImport,
  generateId,
  getBearerToken,
  getMessageBytes,
  initializeRoomCache,
  isDevMode,
  logger,
  matchMaker,
  registerRoomDefinitions,
  requestFromIPC,
  room,
  setDevMode,
  spliceOne,
  subscribeIPC,
  subscribeLobby,
  toNodeHandler,
  unregisterRoomDefinitions,
  updateLobby,
  validate
};
