// @colyseus/core/build/router/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/core/src/router/index.ts
var router_exports = {};
__export(router_exports, {
  __globalEndpoints: () => __globalEndpoints,
  bindRouterToTransport: () => bindRouterToTransport,
  createEndpoint: () => import_better_call2.createEndpoint,
  createInternalContext: () => import_better_call2.createInternalContext,
  createMiddleware: () => import_better_call2.createMiddleware,
  createRouter: () => createRouter,
  toNodeHandler: () => import_node.toNodeHandler
});
module.exports = __toCommonJS(router_exports);
var import_better_call = require("@colyseus/better-call");
var import_node = require("@colyseus/better-call/node");
var import_Transport = require("../Transport.cjs");
var import_controller = require("../matchmaker/controller.cjs");
var import_package = __toESM(require("../../package.json"), 1);
var import_better_call2 = require("@colyseus/better-call");
function bindRouterToTransport(transport, router, useExpress) {
  router.addEndpoint((0, import_better_call.createEndpoint)("/__healthcheck", { method: "GET" }, async (ctx) => {
    return new Response("OK", { status: 200 });
  }));
  const server = transport.server;
  const expressApp = useExpress ? transport.getExpressApp() : server?.listeners("request").find((listener) => listener.name === "app" && listener["mountpath"] === "/");
  const hasRootRoute = (
    // check if express app has a root route
    expressApp && expressRootRoute(expressApp) !== void 0 || // check if router has a root route
    Object.values(router.endpoints).some((endpoint) => endpoint.path === "/")
  );
  if (!hasRootRoute) {
    router.addEndpoint((0, import_better_call.createEndpoint)("/", { method: "GET" }, async (ctx) => {
      return new Response(`Colyseus ${import_package.default.version}`, { status: 200 });
    }));
  }
  if (!server && transport.bindRouter) {
    transport.bindRouter(router);
    return;
  }
  let next;
  if (expressApp) {
    server.removeListener("request", expressApp);
    next = async (req, res) => {
      if (router.findRoute(req.method, req.url.split("?")[0]) !== void 0) {
        const protocol = req.headers["x-forwarded-proto"] || (req.socket.encrypted ? "https" : "http");
        const base = `${protocol}://${req.headers[":authority"] || req.headers.host}`;
        const response = await router.handler((0, import_node.getRequest)({ base, request: req }));
        return (0, import_node.setResponse)(res, response);
      } else {
        return expressApp["handle"](req, res);
      }
    };
  } else {
    next = (0, import_node.toNodeHandler)(router.handler);
  }
  server.prependListener("request", (req, res) => {
    const corsHeaders = {
      ...import_controller.controller.DEFAULT_CORS_HEADERS,
      ...import_controller.controller.getCorsHeaders(new Headers(req.headers))
    };
    if (req.method === "OPTIONS") {
      res.writeHead(204, corsHeaders);
      res.end();
      return;
    }
    Object.entries(corsHeaders).forEach(([key, value]) => {
      res.setHeader(key, value);
    });
    next(req, res);
  });
}
function expressRootRoute(expressApp) {
  const stack = expressApp?._router?.stack ?? expressApp?.router?.stack;
  if (!stack) {
    return false;
  }
  return stack.find((layer) => layer.match("/") && !["query", "expressInit"].includes(layer.name));
}
var __globalEndpoints = {};
function createRouter(endpoints, config = {}) {
  __globalEndpoints = endpoints;
  return (0, import_better_call.createRouter)({ ...endpoints }, config);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  __globalEndpoints,
  bindRouterToTransport,
  createEndpoint,
  createInternalContext,
  createMiddleware,
  createRouter,
  toNodeHandler
});
