// @colyseus/core/build/router/node.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/core/src/router/node.ts
var node_exports = {};
__export(node_exports, {
  createNodeMatchmakingMiddleware: () => createNodeMatchmakingMiddleware
});
module.exports = __toCommonJS(node_exports);
var import_url = require("url");
var matchMaker = __toESM(require("../MatchMaker.cjs"), 1);
var import_node = require("@colyseus/better-call/node");
var import_default_routes = require("./default_routes.cjs");
function readBody(req) {
  return new Promise((resolve, reject) => {
    let data = "";
    req.on("data", (chunk) => {
      data += chunk.toString();
    });
    req.on("end", () => resolve(data ? JSON.parse(data) : {}));
    req.on("error", reject);
  });
}
function getCorsHeaders(req, headers) {
  return {
    ...matchMaker.controller.DEFAULT_CORS_HEADERS,
    ...matchMaker.controller.getCorsHeaders(headers)
  };
}
function createNodeMatchmakingMiddleware() {
  return async (req, res, next) => {
    const url = new import_url.URL(req.url || "/", "http://localhost");
    const isMatchmakeRoute = url.pathname.startsWith(`/${matchMaker.controller.matchmakeRoute}/`);
    if (!isMatchmakeRoute) {
      next();
      return;
    }
    const headers = new Headers(req.headers);
    const corsHeaders = getCorsHeaders(req, headers);
    if (req.method === "OPTIONS") {
      res.writeHead(204, corsHeaders);
      res.end();
      return;
    }
    if (req.method !== "POST") {
      next();
      return;
    }
    const match = url.pathname.match(/^\/matchmake\/(\w+)\/(.+)/);
    if (!match) {
      next();
      return;
    }
    const [, method, roomName] = match;
    try {
      const response = await (0, import_default_routes.postMatchmakeMethod)({
        params: { method, roomName },
        body: await readBody(req),
        headers: req.headers,
        request: { headers },
        asResponse: true
      });
      await (0, import_node.setResponse)(res, response);
    } catch {
      next();
    }
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createNodeMatchmakingMiddleware
});
