// @colyseus/playground/build/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src-backend/index.ts
var src_backend_exports = {};
__export(src_backend_exports, {
  playground: () => playground
});
module.exports = __toCommonJS(src_backend_exports);
var import_path = __toESM(require("path"));
var import_express = __toESM(require("express"));
var import_auth = require("@colyseus/auth");
var import_core = require("@colyseus/core");
var import_colyseus_ext = require("./colyseus.ext.cjs");
function playground() {
  (0, import_colyseus_ext.applyMonkeyPatch)();
  const router = import_express.default.Router();
  router.use("/", import_express.default.static(import_path.default.resolve(__dirname, "..", "build")));
  router.get("/rooms", async (req, res) => {
    const rooms = await import_core.matchMaker.driver.query({});
    const roomsByType = {};
    const roomsById = {};
    rooms.forEach((room) => {
      if (!roomsByType[room.name]) {
        roomsByType[room.name] = 0;
      }
      roomsByType[room.name]++;
      roomsById[room.roomId] = room;
    });
    res.json({
      rooms: Object.keys(import_core.matchMaker.getAllHandlers()),
      roomsByType,
      roomsById,
      auth: {
        // list of OAuth providers
        oauth: Object.keys(import_auth.auth.oauth.providers),
        register: typeof import_auth.auth.settings.onRegisterWithEmailAndPassword === "function",
        anonymous: typeof import_auth.JWT.settings.secret === "string"
      }
    });
  });
  router.get("/__apidocs", async (_, res) => {
    let z = void 0;
    try {
      z = await import("zod");
    } catch (e) {
    }
    res.json(Object.values(import_core.__globalEndpoints).map((endpoint) => {
      var _a, _b;
      return {
        method: endpoint.options.method,
        path: endpoint.path,
        body: z && endpoint.options.body && z.toJSONSchema(endpoint.options.body),
        query: z && endpoint.options.query && z.toJSONSchema(endpoint.options.query),
        metadata: endpoint.options.metadata,
        description: (_b = (_a = endpoint.options.metadata) == null ? void 0 : _a.openapi) == null ? void 0 : _b.description
      };
    }));
  });
  return router;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  playground
});
