// @colyseus/playground/build/index.mjs
// src-backend/index.ts
import path from "path";
import express from "express";
import { auth, JWT } from "@colyseus/auth";
import { matchMaker, __globalEndpoints } from "@colyseus/core";
import { applyMonkeyPatch } from "./colyseus.ext.mjs";
import { fileURLToPath } from "url";
function playground() {
  applyMonkeyPatch();
  const router = express.Router();
  router.use("/", express.static(path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..", "build")));
  router.get("/rooms", async (req, res) => {
    const rooms = await matchMaker.driver.query({});
    const roomsByType = {};
    const roomsById = {};
    rooms.forEach((room) => {
      if (!roomsByType[room.name]) {
        roomsByType[room.name] = 0;
      }
      roomsByType[room.name]++;
      roomsById[room.roomId] = room;
    });
    res.json({
      rooms: Object.keys(matchMaker.getAllHandlers()),
      roomsByType,
      roomsById,
      auth: {
        // list of OAuth providers
        oauth: Object.keys(auth.oauth.providers),
        register: typeof auth.settings.onRegisterWithEmailAndPassword === "function",
        anonymous: typeof JWT.settings.secret === "string"
      }
    });
  });
  router.get("/__apidocs", async (_, res) => {
    let z = void 0;
    try {
      z = await import("zod");
    } catch (e) {
    }
    res.json(Object.values(__globalEndpoints).map((endpoint) => {
      return {
        method: endpoint.options.method,
        path: endpoint.path,
        body: z && endpoint.options.body && z.toJSONSchema(endpoint.options.body),
        query: z && endpoint.options.query && z.toJSONSchema(endpoint.options.query),
        metadata: endpoint.options.metadata,
        description: endpoint.options.metadata?.openapi?.description
      };
    }));
  });
  return router;
}
export {
  playground
};
