// @colyseus/auth/build/JWT.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/auth/src/JWT.ts
var JWT_exports = {};
__export(JWT_exports, {
  JWT: () => JWT
});
module.exports = __toCommonJS(JWT_exports);
var import_jsonwebtoken = __toESM(require("jsonwebtoken"), 1);
var import_express_jwt = require("express-jwt");
var import_better_call = require("@colyseus/better-call");
var JWT = {
  settings: {
    /**
     * The secret used to sign and verify the JWTs.
     */
    secret: void 0,
    verify: {
      /**
       * The first algorithm in the list is used to sign new tokens.
       */
      algorithms: ["HS256"]
    }
  },
  sign: function(payload, options = {}) {
    return new Promise((resolve, reject) => {
      if (options.algorithm === void 0) {
        options.algorithm = JWT.settings.verify.algorithms[0];
      }
      import_jsonwebtoken.default.sign(payload, getJWTSecret(), options, (err, token) => {
        if (err) reject(err.message);
        resolve(token);
      });
    });
  },
  verify: function(token, options) {
    return new Promise((resolve, reject) => {
      import_jsonwebtoken.default.verify(token, getJWTSecret(), options || JWT.settings.verify, function(err, decoded) {
        if (err) reject(err);
        resolve(decoded);
      });
    });
  },
  /**
   * Returns the decoded payload without verifying if the signature is valid
   */
  decode: import_jsonwebtoken.default.decode,
  /**
   * Middleware that verifies JsonWebTokens.
   * Works with both Express and better-call.
   *
   * Example (express):
   *   app.get("/protected_route", auth.middleware(), (req, res) => { ... });
   *
   * Example (better-call):
   *   const protectedRoute = createEndpoint("/protected-route", {
   *     method: "GET",
   *     use: [auth.middleware()],
   *   }, async (ctx) => {
   *     // ctx.context.auth contains the decoded JWT payload
   *   });
   */
  middleware: function(options) {
    const expressjwtMiddleware = (0, import_express_jwt.expressjwt)(Object.assign({
      secret: getJWTSecret(),
      algorithms: JWT.settings.verify.algorithms,
      ...JWT.settings.verify
    }, options));
    const betterCallMiddleware = (0, import_better_call.createMiddleware)(async (ctx) => {
      const authHeader = ctx.getHeader("authorization");
      if (!authHeader) {
        throw new import_better_call.APIError(401, { message: "No authorization header" });
      }
      const [scheme, token] = authHeader.split(" ");
      if (scheme?.toLowerCase() !== "bearer" || !token) {
        throw new import_better_call.APIError(401, { message: "Invalid authorization header format" });
      }
      try {
        const decoded = await JWT.verify(token, options);
        return { auth: decoded };
      } catch (err) {
        throw new import_better_call.APIError(401, { message: err.message || "Invalid token" });
      }
    });
    const middleware = function(reqOrCtx, res, next) {
      if (arguments.length === 3) {
        return expressjwtMiddleware(reqOrCtx, res, next);
      } else {
        return betterCallMiddleware(reqOrCtx);
      }
    };
    middleware.options = betterCallMiddleware.options;
    return middleware;
  }
};
function getJWTSecret() {
  JWT.settings.secret ||= process.env.JWT_SECRET;
  if (!JWT.settings.secret) {
    console.error("\u274C Please provide 'JWT_SECRET' environment variable, or set 'JWT.settings.secret'.");
  }
  return JWT.settings.secret;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  JWT
});
