// @colyseus/auth/build/JWT.mjs
// packages/auth/src/JWT.ts
import jsonwebtoken from "jsonwebtoken";
import { expressjwt } from "express-jwt";
import { APIError, createMiddleware } from "@colyseus/better-call";
var JWT = {
  settings: {
    /**
     * The secret used to sign and verify the JWTs.
     */
    secret: void 0,
    verify: {
      /**
       * The first algorithm in the list is used to sign new tokens.
       */
      algorithms: ["HS256"]
    }
  },
  sign: function(payload, options = {}) {
    return new Promise((resolve, reject) => {
      if (options.algorithm === void 0) {
        options.algorithm = JWT.settings.verify.algorithms[0];
      }
      jsonwebtoken.sign(payload, getJWTSecret(), options, (err, token) => {
        if (err) reject(err.message);
        resolve(token);
      });
    });
  },
  verify: function(token, options) {
    return new Promise((resolve, reject) => {
      jsonwebtoken.verify(token, getJWTSecret(), options || JWT.settings.verify, function(err, decoded) {
        if (err) reject(err);
        resolve(decoded);
      });
    });
  },
  /**
   * Returns the decoded payload without verifying if the signature is valid
   */
  decode: jsonwebtoken.decode,
  /**
   * Middleware that verifies JsonWebTokens.
   * Works with both Express and better-call.
   *
   * Example (express):
   *   app.get("/protected_route", auth.middleware(), (req, res) => { ... });
   *
   * Example (better-call):
   *   const protectedRoute = createEndpoint("/protected-route", {
   *     method: "GET",
   *     use: [auth.middleware()],
   *   }, async (ctx) => {
   *     // ctx.context.auth contains the decoded JWT payload
   *   });
   */
  middleware: function(options) {
    const expressjwtMiddleware = expressjwt(Object.assign({
      secret: getJWTSecret(),
      algorithms: JWT.settings.verify.algorithms,
      ...JWT.settings.verify
    }, options));
    const betterCallMiddleware = createMiddleware(async (ctx) => {
      const authHeader = ctx.getHeader("authorization");
      if (!authHeader) {
        throw new APIError(401, { message: "No authorization header" });
      }
      const [scheme, token] = authHeader.split(" ");
      if (scheme?.toLowerCase() !== "bearer" || !token) {
        throw new APIError(401, { message: "Invalid authorization header format" });
      }
      try {
        const decoded = await JWT.verify(token, options);
        return { auth: decoded };
      } catch (err) {
        throw new APIError(401, { message: err.message || "Invalid token" });
      }
    });
    const middleware = function(reqOrCtx, res, next) {
      if (arguments.length === 3) {
        return expressjwtMiddleware(reqOrCtx, res, next);
      } else {
        return betterCallMiddleware(reqOrCtx);
      }
    };
    middleware.options = betterCallMiddleware.options;
    return middleware;
  }
};
function getJWTSecret() {
  JWT.settings.secret ||= process.env.JWT_SECRET;
  if (!JWT.settings.secret) {
    console.error("\u274C Please provide 'JWT_SECRET' environment variable, or set 'JWT.settings.secret'.");
  }
  return JWT.settings.secret;
}
export {
  JWT
};
