// @colyseus/core/build/matchmaker/controller.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/core/src/matchmaker/controller.ts
var controller_exports = {};
__export(controller_exports, {
  controller: () => controller
});
module.exports = __toCommonJS(controller_exports);
var import_shared_types = require("@colyseus/shared-types");
var import_ServerError = require("../errors/ServerError.cjs");
var import_Debug = require("../Debug.cjs");
var matchMaker = __toESM(require("../MatchMaker.cjs"), 1);
var controller = {
  DEFAULT_CORS_HEADERS: {
    "Access-Control-Allow-Headers": "Origin, X-Requested-With, Content-Type, Accept, Authorization",
    "Access-Control-Allow-Methods": "GET,HEAD,PUT,PATCH,POST,DELETE",
    "Access-Control-Allow-Credentials": "true",
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Max-Age": "2592000"
    // ...
  },
  exposedMethods: ["joinOrCreate", "create", "join", "joinById", "reconnect"],
  allowedRoomNameChars: /([a-zA-Z_\-0-9]+)/gi,
  matchmakeRoute: "matchmake",
  /**
   * You can manually change the default corsHeaders by overwriting the `getCorsHeaders()` method:
   *    ```
   *    import { matchMaker } from "@colyseus/core";
   *    matchMaker.controller.getCorsHeaders = function(headers) {
   *      if (headers.get('referer') !== "xxx") {
   *      }
   *
   *      return {
   *        'Access-Control-Allow-Origin': 'safedomain.com',
   *      }
   *    }
   *    ```
   */
  getCorsHeaders(headers) {
    return {
      ["Access-Control-Allow-Origin"]: headers.get("origin") || "*"
    };
  },
  async invokeMethod(method, roomName, clientOptions = {}, authOptions) {
    if (this.exposedMethods.indexOf(method) === -1) {
      throw new import_ServerError.ServerError(import_shared_types.ErrorCode.MATCHMAKE_NO_HANDLER, `invalid method "${method}"`);
    }
    try {
      return await matchMaker[method](roomName, clientOptions, authOptions);
    } catch (e) {
      (0, import_Debug.debugError)(e);
      throw new import_ServerError.ServerError(e.code || import_shared_types.ErrorCode.MATCHMAKE_UNHANDLED, e.message);
    }
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  controller
});
