// @colyseus/core/build/matchmaker/driver.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/core/src/matchmaker/driver.ts
var driver_exports = {};
__export(driver_exports, {
  getLockId: () => getLockId,
  initializeRoomCache: () => initializeRoomCache
});
module.exports = __toCommonJS(driver_exports);
function getLockId(filterOptions) {
  return Object.keys(filterOptions).map((key) => `${key}:${filterOptions[key]}`).join("-");
}
function initializeRoomCache(initialValues = {}) {
  return {
    clients: 0,
    maxClients: Infinity,
    locked: false,
    private: false,
    metadata: void 0,
    createdAt: initialValues && initialValues.createdAt ? new Date(initialValues.createdAt) : /* @__PURE__ */ new Date(),
    unlisted: false,
    ...initialValues
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getLockId,
  initializeRoomCache
});
