// @colyseus/core/build/matchmaker/RegisteredHandler.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/core/src/matchmaker/RegisteredHandler.ts
var RegisteredHandler_exports = {};
__export(RegisteredHandler_exports, {
  INVALID_OPTION_KEYS: () => INVALID_OPTION_KEYS,
  RegisteredHandler: () => RegisteredHandler
});
module.exports = __toCommonJS(RegisteredHandler_exports);
var import_events = require("events");
var import_Logger = require("../Logger.cjs");
var import_Room = require("./../Room.cjs");
var import_Lobby = require("./Lobby.cjs");
var INVALID_OPTION_KEYS = [
  "clients",
  "locked",
  "private",
  // 'maxClients', - maxClients can be useful as filter options
  "metadata",
  "name",
  "processId",
  "roomId"
];
var RegisteredHandler = class extends import_events.EventEmitter {
  constructor(klass, options) {
    super();
    this.filterOptions = [];
    this.realtimeListingEnabled = false;
    this.klass = klass;
    this.options = options;
    if (typeof klass !== "function") {
      import_Logger.logger.debug("You are likely not importing your room class correctly.");
      throw new Error(`class is expected but ${typeof klass} was provided.`);
    }
  }
  enableRealtimeListing() {
    this.realtimeListingEnabled = true;
    this.on("create", (room) => (0, import_Lobby.updateLobby)(room));
    this.on("lock", (room) => (0, import_Lobby.updateLobby)(room));
    this.on("unlock", (room) => (0, import_Lobby.updateLobby)(room));
    this.on("join", (room) => (0, import_Lobby.updateLobby)(room));
    this.on("leave", (room, _, willDispose) => {
      if (!willDispose) {
        (0, import_Lobby.updateLobby)(room);
      }
    });
    this.on("visibility-change", (room, isVisible) => (0, import_Lobby.updateLobby)(room, isVisible));
    this.on("metadata-change", (room) => (0, import_Lobby.updateLobby)(room));
    this.on("dispose", (room) => (0, import_Lobby.updateLobby)(room, true));
    return this;
  }
  /**
   * Define which fields should be used for filtering rooms.
   * Supports both onCreate options and metadata fields using dot notation.
   *
   * @example
   * // Filter by IRoomCache fields
   * .filterBy(['maxClients'])
   *
   * @example
   * // Filter by metadata fields
   * .filterBy(['difficulty', 'metadata.region'])
   *
   * @example
   * // Mix both
   * .filterBy(['mode', 'difficulty', 'maxClients'])
   */
  filterBy(options) {
    this.filterOptions = options;
    return this;
  }
  /**
   * Define how rooms should be sorted when querying.
   * Supports both room cache fields and metadata fields using dot notation.
   *
   * @example
   * // Sort by number of clients (descending)
   * .sortBy({ clients: -1 })
   *
   * @example
   * // Sort by metadata field
   * .sortBy({ 'metadata.rating': -1 })
   *
   * @example
   * // Multiple sort criteria
   * .sortBy({ 'metadata.skillLevel': 1, clients: -1 })
   */
  sortBy(options) {
    this.sortOptions = options;
    return this;
  }
  getMetadataFromOptions(options) {
    const metadata = this.getFilterOptions(options);
    if (this.sortOptions) {
      for (const field in this.sortOptions) {
        if (field in options && !(field in metadata)) {
          metadata[field] = options[field];
        }
      }
    }
    return Object.keys(metadata).length > 0 ? { metadata } : {};
  }
  /**
   * Extract filter options from client options.
   */
  getFilterOptions(options) {
    return this.filterOptions.reduce((prev, curr, i, arr) => {
      const field = String(arr[i]);
      if (options.hasOwnProperty(field)) {
        if (INVALID_OPTION_KEYS.indexOf(field) !== -1) {
          import_Logger.logger.warn(`option "${field}" has internal usage and is going to be ignored.`);
        } else {
          prev[field] = options[field];
        }
      }
      return prev;
    }, {});
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  INVALID_OPTION_KEYS,
  RegisteredHandler
});
