// @colyseus/core/build/matchmaker/RegisteredHandler.mjs
// packages/core/src/matchmaker/RegisteredHandler.ts
import { EventEmitter } from "events";
import { logger } from "../Logger.mjs";
import "./../Room.mjs";
import { updateLobby } from "./Lobby.mjs";
var INVALID_OPTION_KEYS = [
  "clients",
  "locked",
  "private",
  // 'maxClients', - maxClients can be useful as filter options
  "metadata",
  "name",
  "processId",
  "roomId"
];
var RegisteredHandler = class extends EventEmitter {
  constructor(klass, options) {
    super();
    this.filterOptions = [];
    this.realtimeListingEnabled = false;
    this.klass = klass;
    this.options = options;
    if (typeof klass !== "function") {
      logger.debug("You are likely not importing your room class correctly.");
      throw new Error(`class is expected but ${typeof klass} was provided.`);
    }
  }
  enableRealtimeListing() {
    this.realtimeListingEnabled = true;
    this.on("create", (room) => updateLobby(room));
    this.on("lock", (room) => updateLobby(room));
    this.on("unlock", (room) => updateLobby(room));
    this.on("join", (room) => updateLobby(room));
    this.on("leave", (room, _, willDispose) => {
      if (!willDispose) {
        updateLobby(room);
      }
    });
    this.on("visibility-change", (room, isVisible) => updateLobby(room, isVisible));
    this.on("metadata-change", (room) => updateLobby(room));
    this.on("dispose", (room) => updateLobby(room, true));
    return this;
  }
  /**
   * Define which fields should be used for filtering rooms.
   * Supports both onCreate options and metadata fields using dot notation.
   *
   * @example
   * // Filter by IRoomCache fields
   * .filterBy(['maxClients'])
   *
   * @example
   * // Filter by metadata fields
   * .filterBy(['difficulty', 'metadata.region'])
   *
   * @example
   * // Mix both
   * .filterBy(['mode', 'difficulty', 'maxClients'])
   */
  filterBy(options) {
    this.filterOptions = options;
    return this;
  }
  /**
   * Define how rooms should be sorted when querying.
   * Supports both room cache fields and metadata fields using dot notation.
   *
   * @example
   * // Sort by number of clients (descending)
   * .sortBy({ clients: -1 })
   *
   * @example
   * // Sort by metadata field
   * .sortBy({ 'metadata.rating': -1 })
   *
   * @example
   * // Multiple sort criteria
   * .sortBy({ 'metadata.skillLevel': 1, clients: -1 })
   */
  sortBy(options) {
    this.sortOptions = options;
    return this;
  }
  getMetadataFromOptions(options) {
    const metadata = this.getFilterOptions(options);
    if (this.sortOptions) {
      for (const field in this.sortOptions) {
        if (field in options && !(field in metadata)) {
          metadata[field] = options[field];
        }
      }
    }
    return Object.keys(metadata).length > 0 ? { metadata } : {};
  }
  /**
   * Extract filter options from client options.
   */
  getFilterOptions(options) {
    return this.filterOptions.reduce((prev, curr, i, arr) => {
      const field = String(arr[i]);
      if (options.hasOwnProperty(field)) {
        if (INVALID_OPTION_KEYS.indexOf(field) !== -1) {
          logger.warn(`option "${field}" has internal usage and is going to be ignored.`);
        } else {
          prev[field] = options[field];
        }
      }
      return prev;
    }, {});
  }
};
export {
  INVALID_OPTION_KEYS,
  RegisteredHandler
};
